File: src/25-merge-sort/py/__main__.py
from Merge import merge

from MergeSort import merge_sort

# Tests for merge routine

assert(merge([], []) == [])

assert(merge([1], []) == [1])

assert(merge([], [1]) == [1])

assert(merge([1], [1]) == [1,1])

assert(merge([1,2], [3]) == [1,2,3])

assert(merge([3], [1,2]) == [1,2,3])

assert(merge([1,2,3,4,4,5,6,7,7,8,9], [1,1,5,5,5,5,5,6,6,9]) == [1,1,1,2,3,4,4,5,5,5,5,5,5,6,6,6,7,7,8,9,9])

# Tests for merge_sort routine

assert(merge_sort([]) == [])

assert(merge_sort([5]) == [5])

assert(merge_sort([5,1]) == [1,5])

assert(merge_sort([5,8,1,2,7,9,4,2,1,5,1]) == [1,1,1,2,2,4,5,5,7,8,9])

File: src/25-merge-sort/cpp/main.cpp
#include <vector>
#include <iostream>

#include "../../utils/cpp/Assert.h"

#include "../../utils/cpp/PrintVector.h"
#include "./MergeSort.h"
#include "./Merge.h"

typedef std::vector<int> List;

int main(int argc, char** argv)
{

	assert("Merging two empty vectors yields an empty vector.",
		merge(List({}), List({})) ==
			List({}));

	assert("Merging two one-element vectors yields a sorted pair.",
		merge(List({94}), List({72})) ==
			List({72, 94}));

	assert("We can merge two sorted lists of arbitrary lengths.",
		merge(List({1,4,7}), List({0,3,9})) ==
		List({0,1,3,4,7,9}));

	assert("We can merge two sorted lists of different lengths.",
		merge(List({1,4}), List({3})) ==
		List({1,3,4}));

	assert("mergeSort yields a sorted list.",
		mergeSort(List({5,3})) ==
		List({3,5}));

	assert("mergeSort yields a sorted list.",
		mergeSort(List({3,5,1,9,7,5,9,2,3,8,4,0,1,2,3})) ==
		List({0,1,1,2,2,3,3,3,4,5,5,7,8,9,9}));

	return 0;

}
